drv/lpc55-update-server/src/main.rs

```diff
@@ -279,31 +277,133 @@ fn validate_header_block(
-        )
+    let page_num = block_num as u32; // TODO why not use the same type
+
+    let flash = unsafe { &*lpc55_pac::FLASH::ptr() };
+    let mut flash = drv_lpc55_flash::Flash::new(flash);
+
+    // Can only update opposite image
+    if same_image(img) {
+        return Err(UpdateError::RunningImage);
+    }
+
+    let write_addr = match target_addr(img, page_num) {
+        Some(addr) => addr,
+        None => return Err(UpdateError::OutOfBounds),
     };
 
-    match result {
-        HypoStatus::Success => Ok(()),
-        HypoStatus::OutOfBounds => Err(UpdateError::OutOfBounds),
-        HypoStatus::RunningImage => Err(UpdateError::RunningImage),
-        // Should probably encode the LPC55 flash status into the update
-        // error for good measure but that takes effort...
-        HypoStatus::FlashError(_) => Err(UpdateError::FlashError),
+    // Step one: erase the page. 
+
+    // write_addr is a byte address; page_num is a page index; but the actual
+    // erase machinery operates in terms of flash words. A flash word is 16
+    // bytes in length. So, we need to convert.
+    //
+    // Note that the range used here is _inclusive._
+    //
+    // We wind up needing this in u32 form a lot, so let's cast once and reuse.
+    const WORDSZ: u32 = BYTES_PER_FLASH_WORD as u32;
+    flash.start_erase_range(write_addr / WORDSZ ..= (write_addr + BLOCK_SIZE_BYTES as u32 - 1) / WORDSZ);
+    wait_for_erase_or_program(&mut flash);
+
+    // Transfer each 16-byte flash word (page row) into the write registers in
+    // the flash controller.
+    for (i, row) in flash_page.chunks_exact(BYTES_PER_FLASH_WORD).enumerate() {
+        // TODO: this will be unnecessary if array_chunks stabilizes
+        let row: &[u8; BYTES_PER_FLASH_WORD] = row.try_into().unwrap_lite();
+
+        flash.start_write_row(i as u32, row);
+        while !flash.poll_write_result() {
+            // spin - supposed to be very quick in hardware.
+        }
+    }
+
+    // Now, program the whole page into non-volatile storage. This again uses
+    // page indices, requiring a divide-by-16.
+    flash.start_program(write_addr / WORDSZ);
+    wait_for_erase_or_program(&mut flash);
+
+    Ok(())
+}
+
+/// Utility function that does an interrupt-driven poll and sleep while the
+/// flash controller finishes a write or erase.
+fn wait_for_erase_or_program(flash: &mut drv_lpc55_flash::Flash) -> Result<(), UpdateError> {
+    loop {
+        if let Some(result) = flash.poll_erase_or_program_result() {
+            return result.map_err(|_| UpdateError::FlashError);
+        }
+
+        flash.enable_interrupt_sources();
+        sys_irq_control(notifications::FLASH_IRQ_MASK, true);
+        // RECV from the kernel cannot produce an error, so ignore it.
+        let _ = sys_recv_closed(&mut [], notifications::FLASH_IRQ_MASK, TaskId::KERNEL);
+        flash.disable_interrupt_sources();
+    }
+}
+
+fn same_image(which: UpdateTarget) -> bool {
```

Comment:
Note to the reader: the code from this point down is stuff I've factored out of `lib/hypocalls` and `lib/lpc55-flash` so that it's all in one place, can use the same types without conversions, etc. I have added light comments to things that bit me, but otherwise the lack of documentation here is from the original.